### PR TITLE
handle impromptu indexing

### PR DESCRIPTION
**Pull request by edsu7**
To address : https://github.com/icgc-argo-workflows/argo-modules/issues/28
- generalize variables downstream in `germlinevar` and `dnaalnqc`
- disable common sample check in `germlinevar`,`dnaalnqc` and `prealnqc`
- detect index for BAM,CRAM,VCF
- when index not present generate index
- add in file exists checks

Test scenarios:
sanity check for existing schemas:
```
nextflow run ./tests/subworkflows/icgc-argo-workflows/stage_input -entry test_stage_input_local -c ./tests/subworkflows/icgc-argo-workflows/stage_input/nextflow.prealnqc.config  -profile docker,rdpc_qa
nextflow run ./tests/subworkflows/icgc-argo-workflows/stage_input -entry test_stage_input_local -c ./tests/subworkflows/icgc-argo-workflows/stage_input/nextflow.aln.config  -profile docker,rdpc_qa
```

If file is missing index, generate:
```
nextflow run ./tests/subworkflows/icgc-argo-workflows/stage_input -entry test_stage_input_local -c ./tests/subworkflows/icgc-argo-workflows/stage_input/nextflow.dnaalnqc.config  -profile docker,rdpc_qa
```
File with existing index:
```
nextflow run ./tests/subworkflows/icgc-argo-workflows/stage_input -entry test_stage_input_local -c ./tests/subworkflows/icgc-argo-workflows/stage_input/nextflow.germlinevar.config  -profile docker,rdpc_qa
```

### CODE REVIEW

**Review by lindaxiang — CHANGES_REQUESTED** [review]
TSV header in [`prep/sample`](https://github.com/icgc-argo-workflows/argo-modules/blob/main/modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py#L347) also needs to be updated accordingly:
- cram --> bam_cram
- crai --> bai_crai

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 1]
```diff
@@ -334,18 +334,18 @@ def main():
...
+        elif fp.endswith('crai') or fp.endswith('bai'):
+          bai_crai = os.path.join(os.getcwd(), args.outdir, os.path.basename(fp))
+          os.symlink(os.path.abspath(fp), bai_crai)
         else:
           sys.exit("Error: not supported input file format")
       with open(output_sample_sheet, 'w', newline='') as f:
         csvwriter = csv.writer(f, delimiter=',')
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','cram','crai',"genome_build",'experiment', 'analysis_json'])
-        csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, cram, crai, genome_build,experiment, metadata_json])
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','bam_cram','bai_crai',"genome_build",'experiment', 'analysis_json'])
```
Comment (edsu7) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:347]: Addressing:
https://github.com/icgc-argo-workflows/argo-modules/pull/33#discussion_r1556180277

File: subworkflows/icgc-argo-workflows/stage_input/main.nf [hunk 2]
```diff
@@ -143,41 +145,87 @@ workflow STAGE_INPUT {
...
+    CRAM_INDEX(ch_index_split.cram_to_index)
+    TABIX_TABIX(ch_index_split.vcf_to_index)
+
+
+    //Combine BAM and BAI into single channel
+    ch_index_split.bam_to_index.join(BAM_INDEX.out.bai).map{
+      meta,bam,index ->
+      tuple([meta, bam, index])
+    }.set{indexed_bam}
+
```
Comment (lindaxiang) [inline subworkflows/icgc-argo-workflows/stage_input/main.nf:191]: Is the `map` operator still needed in this case? :-)

**Review by lindaxiang — APPROVED** [review]
Thinking that the `map` operator is not needed when join the channel.

**Pull request merged**